src/lib/operators/insert.cpp

```diff
@@ -107,136 +93,132 @@ std::shared_ptr<const Table> Insert::_on_execute(std::shared_ptr<TransactionCont
-  }
-
-  auto total_rows_to_insert = static_cast<uint32_t>(input_table_left()->row_count());
+  Assert(_target_table->max_chunk_size() > 0, "Expected max chunk size of target table to be greater than zero");
 
-  // First, allocate space for all the rows to insert. Do so while locking the table to prevent multiple threads
-  // modifying the table's size simultaneously.
-  auto start_index = 0u;
-  auto start_chunk_id = ChunkID{0};
-  auto end_chunk_id = 0u;
+  /**
+   * 1. Allocate the required rows in the target Table, without actually copying data to them.
+   *    Do so while locking the table to prevent multiple threads modifying the table's size simultaneously.
+   *    Since allocation is expected to be faster than writing to the memory, allocating under lock and then writing -
+   *    in a second step - without lock will minimize the time that the Table's append_mutex is locked.
+   */
   {
-    auto scoped_lock = _target_table->acquire_append_mutex();
+    const auto append_lock = _target_table->acquire_append_mutex();
+
+    auto remaining_rows = input_table_left()->row_count();
 
     if (_target_table->chunk_count() == 0) {
       _target_table->append_mutable_chunk();
     }
 
-    start_chunk_id = _target_table->chunk_count() - 1;
-    end_chunk_id = start_chunk_id + 1;
-    auto last_chunk = _target_table->get_chunk(start_chunk_id);
-    start_index = last_chunk->size();
+    while (remaining_rows > 0) {
+      auto target_chunk_id = ChunkID{_target_table->chunk_count() - 1};
+      auto target_chunk = _target_table->get_chunk(target_chunk_id);
 
-    // If last chunk is compressed, add a new uncompressed chunk
-    if (!last_chunk->is_mutable()) {
-      _target_table->append_mutable_chunk();
-      end_chunk_id++;
-    }
+      // If the last Chunk of the target Table is either immutable or full, append a new mutable Chunk
+      if (!target_chunk->is_mutable() || target_chunk->size() == _target_table->max_chunk_size()) {
+        _target_table->append_mutable_chunk();
+        ++target_chunk_id;
+        target_chunk = _target_table->get_chunk(target_chunk_id);
+      }
 
-    auto remaining_rows = total_rows_to_insert;
-    while (remaining_rows > 0) {
-      auto current_chunk = _target_table->get_chunk(static_cast<ChunkID>(_target_table->chunk_count() - 1));
-      auto rows_to_insert_this_loop = std::min(_target_table->max_chunk_size() - current_chunk->size(), remaining_rows);
+      const auto target_chunk_num_inserted_rows =
+          std::min<size_t>(_target_table->max_chunk_size() - target_chunk->size(), remaining_rows);
 
-      // Resize MVCC vectors.
-      current_chunk->get_scoped_mvcc_data_lock()->grow_by(rows_to_insert_this_loop, MvccData::MAX_COMMIT_ID);
+      _target_chunk_ranges.emplace_back(
+          ChunkRange{target_chunk_id, target_chunk->size(),
+                     static_cast<ChunkOffset>(target_chunk->size() + target_chunk_num_inserted_rows)});
```

Comment:
Is the end position really necessary?

Reply:
"Not necessary" as in "there are other ways to solve this". "Necessary" as in "this is how I chose to solve this". ;)

Having the end position prevents us from having to check for "have we consumed all input" and "how many rows can we copy into the current chunk" when we actually copy the data in the `2.` step.

Reply:
The discussion about the general structure needs to be resolved first. Generally, I'd like to keep temporary data structures small, especially if we can easily recreate that information later.

Reply:
Remember that you're forcing the reader to remember all the things you put into that struct. If you recreate that information later, the reader does not need as much heap space and the information will be fresh. ;)

Reply:
The readers heap/stack is one of my core concerns, yes. But their compute power is as well. (Re)constructing **where data went/goes** needs to happen in four places: When allocating, when copying, when committing and when rolling back. I'd rather require my reader to have a model of a vector of ranges within a Chunks then having them to parse redundant computations in four different places.

Reply:
Works for me, still wanted to bring it up.

Reply:
This is roughly how on_commit_records will look like with and without explicitly storing the ChunkRanges. This increased complexity would be there for commit_records/rollback_records and the copy step. I think it is desirable to store the ChunkRanges, given the increased complexity incurred without them

```c++

void Insert::_on_commit_records(const CommitID cid) {
  // Without explicitly storing the target ChunkRanges
  for (auto target_chunk_id = _target_table_begin.chunk_id; target_chunk_id < _target_table_end.chunk_id;
       ++target_chunk_id) {
    const auto target_chunk = _target_table->get_chunk(target_chunk_id);

    const auto target_chunk_begin_offset =
        target_chunk_id == _target_table_begin.chunk_id ? _target_table_begin.chunk_offset : ChunkOffset{0};
    const auto target_chunk_end_offset =
        target_chunk_id == _target_table_end.chunk_id ? _target_table_end.chunk_offset : target_chunk->size();

    auto mvcc_data = target_chunk->get_scoped_mvcc_data_lock();

    for (auto chunk_offset = target_chunk_begin_offset; chunk_offset < target_chunk_end_offset; ++chunk_offset) {
      mvcc_data->begin_cids[chunk_offset] = cid;
      mvcc_data->tids[chunk_offset] = 0u;
    }
  }

  // With explicitly storing the target ChunkRanges
  for (const auto& target_chunk_range : _target_chunk_ranges) {
    const auto target_chunk = _target_table->get_chunk(target_chunk_range.chunk_id);
    auto mvcc_data = target_chunk->get_scoped_mvcc_data_lock();

    for (auto chunk_offset = target_chunk_range.begin_chunk_offset; chunk_offset < target_chunk_range.end_chunk_offset;
         ++chunk_offset) {
      mvcc_data->begin_cids[chunk_offset] = cid;
      mvcc_data->tids[chunk_offset] = 0u;
    }
  }
}
```